### DISCUSSION TITLE
How can I run it inside Docker rootless?

### DISCUSSION THREAD

**Discussion by srfsh** [slint-ui/slint#6387] [Q&A] [answered]
Hi,

I often use Docker to containerize everything I would need in a project. This mostly includes the tooling, language servers, and such. I actually never had the opportunity, until now, to test something graphical. I was venturing on this journey of graphical interaction using Docker, and luckily, some folks have pointed out that I can use `xhost` to allow the Docker to use X11 on my host machine in combination with bind-mounting `/tmp/.X11-unix` socket. It appears to be working quite well with apps such as xclock, xlogo, etc., but fails to do so with Slint. I have absolutely no idea why it does so. It just prints out an error like `Error: Other("Could not initialize backend.\nError from Winit backend: Error initializing winit event loop: the requested operation is not supported by Winit\nNo backends configured.")` without any further indication of what might've gone wrong. I can't debug it any further than that. I would greatly appreciate any guidance on how to debug this or basically make it work. :)

Many thanks!

**Comment by tronical** [comment]
I think by default winit will try to connect to the wayland compositor first, and perhaps due to the environment variables in your container, that is attempted but somehow fails, and it never goes to trying the x11 display server. Do you have `WAYLAND_DISPLAY` set by chance?

**Reply by srfsh** [reply]
No, sorry. I am using https://github.com/clux/muslrust for what is worth. I forgot to mention I have probably set my `DISPLAY` for X11, of course. One thing that might be important to mention is my setup with Docker. Here is the Dockerfile I am using to test stuff out:

```
FROM docker.io/clux/muslrust:nightly
RUN rustup component add rust-analyzer rustfmt
RUN RUSTFLAGS="-C target-feature=-crt-static -C link-arg=-lm" \
	CARGO_HOME=/tmp/cargo cargo install slint-lsp \
		&& mv /tmp/cargo/bin/slint-lsp /root/.cargo/bin \
		&& rm -rf $CARGO_HOME
```

**Comment by tronical** [comment]
> I have probably set my DISPLAY for X11, of course

Neither this docker file nor muslrust seem to set any display related variables. How do you set them?

**Reply by srfsh** [reply]
I meant to say "properly set". It is set when I am running it with `docker run --rm -e "DISPLAY=$DISPLAY" ...`

**Reply by tronical** [reply]
Okay, that sounds good. Off the top of my head, I don't know what could be causing this then. Perhaps you could try running it through strace to see which socket it tries to connect to and why perhaps that fails?

**Comment by srfsh** [comment, marked as answer]
Thanks a bunch for mentioning strace. It is strage I didn't think of that. I have updated my Dockerfile and it seems to function well now. I hope somebody could put this in use as well (make sure you mount `/tmp/.X11-unix` and run `xhost +local:docker` on your host machine):

```
FROM docker.io/clux/muslrust:nightly
RUN rustup component add rust-analyzer rustfmt
RUN apt-get update \
	&& apt-get install -y libx11-dev libxcursor-dev libx11-xcb-dev libxi-dev libxkbcommon-x11-dev \
	&& rm -rf /var/lib/apt/lists/*
RUN RUSTFLAGS="-C target-feature=-crt-static -C link-arg=-lm" \
	CARGO_HOME=/tmp/cargo cargo install slint-lsp \
		&& mv /tmp/cargo/bin/slint-lsp /root/.cargo/bin \
```

One thing to note here is that the RUSTFLAGS must be duplicated in .cargo/config.toml for `[build]`:

```
[build]
rustflags = ["-C", "target-feature=-crt-static", "-C", "link-arg=-lm"]
```

Now, all I need is a way to compile stuff for Android, too. Then, I can continue on hacking!
